src/main/java/io/github/syst3ms/skriptparser/sections/SecLoop.java

```diff
@@ -65,54 +86,67 @@ public boolean init(Expression<?>[] expressions, int matchedPattern, ParseContex
 				return false;
 			}
 		}
-		lambda = SkriptConsumer.create(this);
 		return true;
 	}
 
 	@Override
 	public Optional<? extends Statement> walk(TriggerContext ctx) {
-		if (isNumericLoop && expr == null) {
+		if (isNumericLoop && expression == null) {
 			// We just set the looped expression to a range from 1 to the amount of times.
 			// This allows the usage of 'loop-number' to get the current iteration
-			expr = rangeOf(ctx, times);
+			expression = rangeOf(ctx, times);
 		}
 
-		Iterator<?> iter = iterators.get(this);
-		if (iter == null) {
-			iter = expr instanceof Variable ? ((Variable<?>) expr).variablesIterator(ctx) : expr.iterator(ctx);
-			if (iter.hasNext()) {
-				iterators.put(this, iter);
-			} else {
-				iter = null;
-			}
+		if (iterator == null) {
+			assert expression != null;
+			iterator = expression instanceof Variable ? ((Variable<?>) expression).variablesIterator(ctx) : expression.iterator(ctx);
 		}
-		var finalIter = iter;
-		return Optional.ofNullable(iter)
-				.filter(Iterator::hasNext)
-				.map(it -> {
-					lambda.accept(ctx, it.next());
-					return (Statement) this;
-				})
-				.or(() -> {
-					if (finalIter != null)
-						iterators.remove(this);
-					return getNext();
-				});
+
+		if (iterator.hasNext()) {
+			setArguments(iterator.next());
+			return start();
+		} else {
+			finish();
+			return Optional.ofNullable(actualNext);
+		}
+	}
+
+	@Override
+	public Statement setNext(@Nullable Statement next) {
+		this.actualNext = next;
+		return this;
+	}
+
+	@Override
+	public void finish() {
+		// Cache clearing
+		iterator = null;
+	}
+
+	@Override
+	public Optional<? extends Statement> getContinued(TriggerContext ctx) {
+		return Optional.of(this);
+	}
+
+	@Override
+	public Optional<Statement> getActualNext() {
+		return Optional.ofNullable(actualNext);
 	}
 
 	@Override
 	public String toString(TriggerContext ctx, boolean debug) {
-		return "loop " + (isNumericLoop ? times.toString(ctx, debug) + " times" : expr.toString(ctx, debug));
+		assert expression != null;
+		return "loop " + (isNumericLoop ? times.toString(ctx, debug) + " times" : expression.toString(ctx, debug));
 	}
 
 	/**
 	 * @return the expression whose values this loop is iterating over
 	 */
 	public Expression<?> getLoopedExpression() {
-		if (isNumericLoop && expr == null) {
-			expr = rangeOf(TriggerContext.DUMMY, times);
+		if (isNumericLoop && expression == null) {
+			expression = rangeOf(TriggerContext.DUMMY, times);
 		}
-		return expr;
+		return expression;
 	}
 
 	private static Expression<BigInteger> rangeOf(TriggerContext ctx, Expression<BigInteger> size) {
```

Comment:
not on the pr but add javadoc for this method ?

Reply:
Not really sure if this is wanted since it is private, but I guess it couldn't hurt.